File: leetcode/best_time_to_buy_and_sell_stock.py
# You are given an array prices where prices[i] is the price of a given stock on the ith day.

# You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.

# Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.

# min_priceに最小値を代入しておき、その最小値とpriceの差分で一番大きいものを最大利益とする。

from typing import List
import sys


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit = 0
        min_price = sys.maxsize
        for price in prices:
            min_price = min(price, min_price)
            max_profit = max(max_profit, price - min_price)

        return max_profit

File: leetcode/easy/best_time_to_buy_and_sell_stock.ts
function maxProfit(prices: number[]): number {
    let max_profit = 0
    let min_price = Number.MAX_SAFE_INTEGER

    for (let price of prices) {
        min_price = Math.min(min_price, price)
        max_profit = Math.max(max_profit, price - min_price)
    }

    return max_profit
};
